Added tutorial for Github <-> Twilio notifications (#79)
cloudflare/workers-docs-engine: src/content/tutorials/github-twilio-notifications/index.md

- A Wrangler installation running locally on your machine, and access to the command-line

Follow our [Workers Quickstart Guide](https://developers.cloudflare.com/workers/quickstart) to setup your local development environment and create a Cloudflare account.

If you don't have a Github repo or account, you can create one on [Github](wwww.github.com) and [create a new repo](https://docs.github.com/en/github/getting-started-with-github/create-a-repo).

### Generate

Cloudflare’s command-line tool for managing Worker projects, Wrangler, has great support for templates — pre-built collections of code that make it easy to get started writing Workers. In this tutorial, you’ll use the default JavaScript template to generate a Workers project.

In the command line, generate your Worker project, passing in a project name (e.g. “github-sms-notifications”):

```sh
---
header: Generate a project
---
wrangler generate github-sms-notifications
cd todos
```

Wrangler templates are just Git repositories, so if you want to create your own templates, or use one from our [Template Gallery](/templates), there’s a ton of options to help you get started.

> Wrangler’s default template includes support for building and deploying JavaScript-based projects, including Webpack support. Inside of your new `todos` directory, `index.js` represents the entry-point to your Cloudflare Workers application.

Author: :)

```javascript
---
filename: index.js
---
addEventListener("fetch", event => {
  event.respondWith(handleRequest(event.request))
})
/**
 * Fetch and log a request
 * @param {Request} request
 */
async function handleRequest(request) {
  return new Response("Hello worker!", { status: 200 })
}
```

In your default `index.js` file, we can see that request/response pattern in action. The `handleRequest` constructs a new `Response` with the body text “Hello worker”, as well as an explicit status code of 200. When a `fetch` event comes into the worker, the script uses `event.respondWith` to return that new response back to the client. This means that your Cloudflare Worker script will serve new responses directly from Cloudflare's cloud network: instead of continuing to the origin, where a standard server would accept requests, and return responses, Cloudflare Workers allows you to respond quickly and efficiently by constructing responses directly on the edge.

> ---- 

Author: removed

Awesome, you should now be ready to go after finishing the prequisites section.

You can reference the finished code on this [Github repository](https://github.com/davidtsong/github-twilio-notifications/)

> At a high level, we will first create a webhook for a Github repo to post an update to your worker. Then, we will parse this update from Github with your worker. Finally, we will send a POST request to Twilio to send a text message to you.

Author: hahaha smart quotes

filename: index.js
---
addEventListener("fetch", event => {
  event.respondWith(handleRequest(event.request))
})
/**
 * Fetch and log a request
 * @param {Request} request
 */
async function handleRequest(request) {
  return new Response("Hello worker!", { status: 200 })
}
```

Let's start and refactor the starter code to handle a POST response and rename the request handler. We can use the `request.method` property of [request](link to request) to check if the request is a `POST` request and send an error response if incorrect. The `simpleResponse` fucntion is an easy wrapper for you to send requests with your Worker.

```javascript
---
filename: index.js
---
function simpleResponse(statusCode, message) {
  let resp = {
    message: message,
    status: statusCode
  }

  return new Response(JSON.stringify(resp), {
    headers: new Headers([["Content-Type", "application/json"]]),
    status: statusCode
  })
}

addEventListener("fetch", event => {
  event.respondWith(githubWebhookHandler(event.request))
})

async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,
      `Please send a POST request :)`
    )
  }
}
```

> The next challenge is to check validate the request is sent with the right secret key. GitHub attaches a hash signature for each payload using the secret key. [Github's Documentation](https://developer.github.com/webhooks/securing/) We can use a helper function called checkSignature on the request the hash is correct. Then, we can access data from the webhook by parsing the request as json.

Reviewer: ```suggestion
Next, we'll validate that the request is sent with the right secret key. GitHub attaches a hash signature for each payload using the secret key ([see the docs for more details](https://developer.github.com/webhooks/securing/)). We can use a helper function called `checkSignature` on the request to ensure the hash is correct. Then, we can access data from the webhook by parsing the request as json.
```

```javascript
---
filename: index.js - githubWebhookHandler()
---
async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,
      `Please send a POST request :)`
    )
  }

>   try{

Reviewer: ```suggestion
  try {
```

    if (!checkSignature(request)) {
      return simpleResponse(403, "Incorrect Secret Code")
    }

    const formData = await request.json()
    const repo_name = formData.repository.full_name

>     const action = await request.headers.get("X-GitHub-Event")

Reviewer: AFAIK this doesn't needs an `await`, try it w/o and see if it still works

    const sender_name = formData.sender.login
  }
  catch (e) {
    return simpleResponse(
      200,
      `Error:  ${e} `
    )
  }
}
```

> The checkSignature function will use the crypto library to hash the received payload with our known secret key to ensure it matches the request hash. GitHub uses a HMAC hexdigest to compute the hash in the sha1 format [Github's Documentation](https://developer.github.com/webhooks/securing/).

Reviewer: ```suggestion
The `checkSignature` function will use the crypto library to hash the received payload with our known secret key to ensure it matches the request hash. GitHub uses a HMAC hexdigest to compute the hash in the sha1 format [Github's Documentation](https://developer.github.com/webhooks/securing/).
```

Reviewer: you linked to this docs page earlier, should be ok to remove

```javascript
---
filename: index.js
---

const crypto = require('crypto')

async function createHexSignature(requestBody) {
  let hmac = crypto.createHmac('sha1', SECRET_TOKEN)
  hmac.update(requestBody,'utf-8')

  return hmac.digest('hex')
}

async function checkSignature(request) {
  let expectedSignature = await createHexSignature(await request.text())

>   let actualSignature = await request.headers.get("X-Hub-Signature")

Reviewer: same comment re: `await` as above

  return expectedSignature === actualSignature
}
```

> Since our project relies on importing a library, we need to also update our `wrangler.toml` and set `type = "webpack"`.

Reviewer: should have a code sample here, and i'd also considering moving this up into the prereqs/setup section, but up to you on that one

## 3. Sending a Text with Twilio

> Finally, we will send a text message to you with the Github repo update info using Twilio. You'll need a Twilio account and number to send texts so follow [this guide to get setup](https://www.twilio.com/sms/api). FYI, Twilio has this [pretty awesome game](https://www.twilio.com/quest) where you can learn how to use their platform and get some free credits while you are at it.

Reviewer: i really like this part! sounds like it's written by a real person :)

> Then, we can create a helper function to send text messages for us by sending a post request to the Twilio API endpoint. [Refer to the Twilio reference here](https://www.twilio.com/docs/sms/api/message-resource#create-a-message-resource)

Reviewer: ```suggestion
Then, we can create a helper function to send text messages for us by sending a post request to the Twilio API endpoint. [Refer to the Twilio reference here](https://www.twilio.com/docs/sms/api/message-resource#create-a-message-resource) to learn more about this endpoint.
```

> We construct our headers and body in the format show n in the Twilio reference. Change the from value to your Twilio Phone number.

Reviewer: ```suggestion
We construct our headers and body in the format shown in the Twilio reference. Change the `from` value to your Twilio Phone number.
```

```javascript
---
filename: index.js - sendText()
---
async function sendText(message){
  let headers = new Headers();

>   headers.append("Authorization","Basic " + Buffer.from(accountSid + ":" + authToken).toString('base64'))

Reviewer: this is interesting, are these secret values in a format where we need to use the `Buffer` class? can we just do `Basic ${accountSid}:${authToken}`?

  headers.append("Content-Type", "application/x-www-form-urlencoded")

  const endpoint = "https://api.twilio.com/2010-04-01/Accounts/" + accountSid + "/Messages.json"

  let encoded = new URLSearchParams()
  encoded.append("To",recipient)
  encoded.append("From", '+19388887573')
  encoded.append("Body", message)

  const request = {
    body: encoded,
    method: 'POST',

>     headers: headers

Reviewer: generally you can simplify this by passing in a hash of key value pairs, instead of constructing headers at the beginning of the function:

```js
headers: {
  "Authorization": `Basic ${token}`,
  "Content-Type": "application/x-www-form-urlencoded"
}
```

  }

  let result = await fetch(endpoint, request)
  result = await result.json()

  return new Response(JSON.stringify(result), request)
}
```
In order to make this work, we need to set some secrets to hide your account SID and authToken from the source code. You can set secrets as such to get a prompt for the value of the secret name you input:

```sh

> wrangler secret put accountSID

Reviewer: strong preference for us to use constant-style variable naming here: `ACCOUNT_SID` and `AUTH_TOKEN`

```

Use this syntax to set your Twilio accountSID, recipient(your number), Twilio authToken, and SECRET_TOKEN (for Github) to the respecive values.

Finally, we modify our `githubWebhookHandler` to send a text at the end.

```javascript
async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,
      `Please send a POST request :)`
    )
  }
  try {
    if (!checkSignature(request)) {
      return simpleResponse(403, "Incorrect Secret Code")
    }

    const formData = await request.json()
    const repo_name = formData.repository.full_name
    const action = await request.headers.get("X-GitHub-Event")
    const sender_name = formData.sender.login
    
    return await sendText(`${sender_name} performed ${action} on ${repo_name}`)

  } catch (e) {
    return simpleResponse(
      200,
      `Error:  ${e} `
    )
  }
}
```

> Congrats, you finished this tutorial! Test it out and make a push to your repo and you should get a text message soon after. If you've never used Git before, here's a [quick guide](https://www.datacamp.com/community/tutorials/git-push-pull) to pushing to your repo. 

Reviewer: idea that would be awesome but no pressure: a GIF of your phone receiving a text message from this worker :)

Reviewer: also would recommend linking to the final repo again here!

You've learned how to:
- use Workers with webhooks
- Integrate Workers with Github and Twilio
- use Worker secrets with wrangler

Next steps:

> - Auth0 tutorial

Reviewer: flesh these links out!

---

> updated: 2020-06-05

Reviewer 2: Should this be updated to `2020-08-17`?

difficulty: Intermediate
---

> <!-- To Do:

Reviewer 2: `TODO:`

 - fix error coding http 200 / 400 /etc
 - should we add more simpleRequests to a standard lib?
 -->

> # GitHub repo notifications with Twilio

Reviewer 2: The URL path should more closely match the `h1` (or vice-versa).

In this tutorial, you will build an SMS notification system on Workers to receive updates about one of your GitHub repos. Whenever there is new activity on your GitHub repo, your worker will send you a text using Twilio.

You’ll learn how to:
- Build webhooks using Workers
- Integrate Workers with GitHub and Twilio

> - use Worker secrets with wrangler

Reviewer 2: `use` here is lowercased when `Build` and `Integrate` are upper.

> ![Finished](media/notification.gif)

Reviewer 2: Need more descriptive name for image file name and alt text. See [accessibility guidelines](https://workers.cloudflaredocs.workers.dev/docs-engine/contributors-guide#accessibility-guidelines).

## Prerequisites

You will need a Worker and a GitHub repository to get started.

To set up a Worker, you should have a:

- A Cloudflare account, and access to the API keys for that account
- A Wrangler installation running locally on your machine, and access to the command-line

Follow our [Workers Quickstart Guide](https://developers.cloudflare.com/workers/quickstart) to set up your local development environment and create a Cloudflare account.

> If you don't have a GitHub repo or account, you can create one on [GitHub](wwww.GitHub.com) and [create a new repo](https://docs.GitHub.com/en/GitHub/getting-started-with-GitHub/create-a-repo).

Reviewer 2: This is not how we do prerequisites anymore. Please copy the other tutorials which now use a "Before you start" partial, e.g.:

https://github.com/cloudflare/workers-docs-engine/blob/eff19a778867b2d5255b7bf9d16658969e5b6088/src/content/tutorials/configure-your-cdn/index.md#L1-L12

In the command line, generate your Worker project, passing in a project name (e.g. `github-twilio-notifications`):

```sh
---
header: Generate a project
---
wrangler generate github-twilio-notifications
cd github-twilio-notifications
```

Wrangler templates are just Git repositories, so if you want to create your own templates, or use one from our [Template Gallery](/templates), there’s a ton of options to help you get started.

Wrangler’s default template includes support for building and deploying JavaScript-based projects, including Webpack support. Inside of your new `github-sms-notifications` directory, `index.js` represents the entry-point to your Cloudflare Workers application.

Awesome, you are ready to go now after finishing the prerequisites section.

First, we'll create a webhook for your GitHub repo to post updates to your worker. Inside of your worker, we'll parse the updates GitHub. Finally, we'll send a POST request to Twilio to send a text message to you.

> You can reference the finished code on this [GitHub repository](https://github.com/davidtsong/github-twilio-notifications/)

Reviewer 2: Sentence missing period `.` at the end.

> ---- 

Reviewer 2: Use 32: `--------------------------------`.

Also, remove unnecessary whitespace below here.

## Configure GitHub

The first step is to configure a GitHub webhook to post to our Worker when there is an update to the repo.

Click on the "Settings" tab for your repo and then click on the "Webhooks" page from the list on the left. Click "Add webhook" and configure the "Payload URL" with your worker URL. You can find your worker URL by running "wrangler publish" in your command line to generate a live link for your Worker. For the "Content type", choose "application/json". Generate a secret key of your choice and set it as the "Secret". Click "Let me select individual events" for which events to trigger the webhook. Select the events you want to get notifications for. (I selected "Pull requests", "Pushes", and "Branch or tag creation") Finally, click "Add webhook" to finish our work on the GitHub side.

![Your config should look something like this](media/ghconfig.png)

> 

Reviewer 2: Add missing `--------------------------------`.

## Parsing the Response

With your local environment setup, we will now parse the repo update from GitHub with your worker. If you get stuck, you can refer to the finished [index.js](https://GitHub.com/davidtsong/GitHub-twilio-notifications/blob/master/index.js). 

Your generated `index.js` should look like this below:

```javascript
---
filename: index.js
---
addEventListener('fetch', event => {
  event.respondWith(handleRequest(event.request))
})

> /**

Reviewer 2: Missing new line above.

 * Respond with hello worker text
 * @param {Request} request
 */
async function handleRequest(request) {

>   return new Response('Hello worker!', {

Reviewer 2: Use double-quotes `"` for strings in JS code.

    headers: { 'content-type': 'text/plain' },
  })
}
```

Let's start and refactor the starter code to handle a `POST` response and rename the request handler. We can use the `request.method` property of [request](/runtime-apis/request) to check if the request is a `POST` request, and send an error response if incorrect. The `simpleResponse` function is an easy wrapper for you to send requests with your Worker.

```javascript
---
filename: index.js
---
function simpleResponse(statusCode, message) {
  let resp = {
    message: message,
    status: statusCode
  }

  return new Response(JSON.stringify(resp), {
    headers: { "Content-Type": "application/json" },
    status: statusCode
  })
}

addEventListener("fetch", event => {
  event.respondWith(githubWebhookHandler(event.request))
})

async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,

>       `Please send a POST request :)`

Reviewer 2: Use `"` since there are no interpolations `${}`.

    )
  }
}
```

> Next, we'll validate that the request is sent with the right secret key. GitHub attaches a hash signature for each payload using the secret key ([see the docs for more details](https://developer.github.com/webhooks/securing/)). We can use a helper function called `checkSignature` on the request to ensure the hash is correct. Then, we can access data from the webhook by parsing the request as JSON.

Reviewer 2: Use [curly quotes](https://workers.cloudflaredocs.workers.dev/docs-engine/contributors-guide#typography).

```javascript
---
filename: index.js - githubWebhookHandler()
---
async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,
      `Please send a POST request :)`
    )
  }
  try {
    const formData = await request.json()
    const headers = await request.headers
    const action = headers.get("X-GitHub-Event")
    const repo_name = formData.repository.full_name
    const sender_name = formData.sender.login 

    if (!checkSignature(formData, headers)) {
      return simpleResponse(403, "Wrong password, try again :P")
    }

  } catch (e) {
    return simpleResponse(
      200,

>       `Error:  ${e} `

Reviewer 2: Remove unnecessary extra space after `${e}`.

    )
  }
}

```

The `checkSignature` function will use the crypto library to hash the received payload with our known secret key to ensure it matches the request hash. GitHub uses an HMAC hexdigest to compute the hash in the sha1 format.

```javascript
---
filename: index.js
---
const crypto = require('crypto')

async function createHexSignature(requestBody) {
  let hmac = crypto.createHmac('sha1', SECRET_TOKEN)
  hmac.update(requestBody,'utf-8')

  return hmac.digest('hex')
}

async function checkSignature(formData, headers) {
  let expectedSignature = await createHexSignature(formData)
  let actualSignature = headers.get("X-Hub-Signature")

  return expectedSignature === actualSignature
}
```

Since our project relies on importing a library, we need to also update our `wrangler.toml` and set `type = "webpack"`.

> Your `wrangler.toml` shuold look something like this:

Reviewer 2: Add new line below.

> ```yaml

Reviewer 2: Use `toml` here not `yaml`.

Reviewer 2: Also add filename?

```javascript
async function githubWebhookHandler(request) {
  if (request.method !== "POST") {
    return simpleResponse(
      200,
      `Please send a POST request :)`
    )
  }
  try {
    const formData = await request.json()
    const headers = await request.headers
    const action = headers.get("X-GitHub-Event")
    const repo_name = formData.repository.full_name
    const sender_name = formData.sender.login 

    if (!checkSignature(formData, headers)) {
      return simpleResponse(403, "Wrong password, try again :P")
    }
    
    return await sendText(`${sender_name} casted spell: ${action} onto your repo ${repo_name}`)

  } catch (e) {
    return simpleResponse(
      200,
      `Error:  ${e} `
    )
  }
}
```

That's all :) Just use wrangler to deploy your script to Workers.
```sh
wrangler publish
```

![Finished](media/notification.gif)

Congrats, you finished this tutorial! Test it out and make a push to your repo and you should get a text message soon after. If you've never used Git before, here's a [quick guide](https://www.datacamp.com/community/tutorials/git-push-pull) to pushing to your repo. 

> You can reference the finished code on this [GitHub repository](https://github.com/davidtsong/github-twilio-notifications/)

Reviewer 2: Missing `.`.
